File: roktracker/utils/types/full_config.py
from pydantic import BaseModel
from typing import Dict, List, Optional


class TimingsConfig(BaseModel):
    gov_open: float = 2.0
    copy_wait: float = 0.2
    kills_open: float = 1.0
    info_open: float = 1.0
    info_close: float = 0.5
    gov_close: float = 1.0
    max_random: float = 0.5


class FormatsConfig(BaseModel):
    xlsx: bool = True
    csv: bool = False
    jsonl: bool = False

    def from_list(self, list: List[str]):
        for item in list:
            if item == "xlsx":
                self.xlsx = True
            elif item == "csv":
                self.csv = True
            elif item == "jsonl":
                self.jsonl = True

    def from_dict(self, dict: Dict[str, bool]):
        for key, value in dict.items():
            if key == "xlsx":
                self.xlsx = value
            elif key == "csv":
                self.csv = value
            elif key == "jsonl":
                self.jsonl = value


class ScanConfig(BaseModel):
    kingdom_name: str = ""
    people_to_scan: int = 300
    resume: bool = False
    advanced_scroll: bool = True
    track_inactives: bool = False
    validate_power: bool = False
    power_threshold: int = 100000
    validate_kills: bool = True
    reconstruct_kills: bool = True
    timings: TimingsConfig
    formats: FormatsConfig


class BluestacksConfig(BaseModel):
    name: str = "RoK Tracker"
    config: str = "C:\\ProgramData\\BlueStacks_nxt\\bluestacks.conf"


class GeneralConfig(BaseModel):
    emulator: str = "bluestacks"
    bluestacks: BluestacksConfig
    adb_port: int = 5555


class FullConfig(BaseModel):
    scan: ScanConfig
    general: GeneralConfig

File: gui_frontend/src/schema/FullConfig.ts
import { z } from "zod";

export const FullConfigSchema = z.object({
  scan: z.object({
    kingdom_name: z.string().default(""),
    people_to_scan: z.number().int().default(300),
    resume: z.boolean().default(false),
    advanced_scroll: z.boolean().default(true),
    track_inactives: z.boolean().default(false),
    validate_power: z.boolean().default(false),
    power_threshold: z.number().int().default(100000),
    validate_kills: z.boolean().default(true),
    reconstruct_kills: z.boolean().default(true),
    timings: z.object({
      gov_open: z.number().default(2),
      copy_wait: z.number().default(0.2),
      kills_open: z.number().default(1),
      info_open: z.number().default(1),
      info_close: z.number().default(0.5),
      gov_close: z.number().default(1),
      max_random: z.number().default(0.5),
    }),
    formats: z.object({
      xlsx: z.boolean().default(true),
      csv: z.boolean().default(false),
      jsonl: z.boolean().default(false),
    }),
  }),
  general: z.object({
    emulator: z.string().default("bluestacks"),
    bluestacks: z.object({
      name: z.string().default("RoK Tracker"),
      config: z
        .string()
        .default("C:\\ProgramData\\BlueStacks_nxt\\bluestacks.conf"),
    }),
    adb_port: z.number().int().default(5555),
  }),
});
export type FullConfig = z.infer<typeof FullConfigSchema>;
